I20 trajectory 1

yshiyi opened the pull request:
A class of input functions which can generate four particular trajectories such as spiral, nudges, swerves and corner has been defined. 
Four examples have been designed to present the output of these four input functions.
User may check a specific example by choosing the accosiated index number at the beginning of the program.
User may also give an arbitrary initial condition to the motion model right after choosing the index number.
In addition, a new motion model module has been constructed so that users may generate any motions by modifying this module.

yshiyi committed:
New trajectory generation module

Allow user to define any motion model in motion_model.m.
In robottraj_new.m, user may choose a trajectory from pre-defined ones: swerves, nudges, lane change, corner.
User may also give any initial condition to generate a trajectory.

yshiyi committed:
Backup of the original robottraj

yshiyi committed:
This is the new robottraj

Users may choose a trajectory from pre-defined ones: swerves, nudge, lane change, corner.
Users may also choose the motion model which they have defined  in motion_model.m.

yshiyi committed:
This is the updated robottraj file.

Add one new trajectory spiral and combine swerves and lane change together.

yshiyi committed:
Delete robottraj_new.m

yshiyi committed:
Update robottraj.m

Add some comments

yshiyi committed:
Update motion_model.m

Add some commends

chutsu commented:
Is `i20_Trajectory` not the one you are editing? if not please remove it?

chutsu commented:
Several points to consider:
- Maybe put this in the `matlab_simulation/7-Control/` folder?
- Since you are asking the user for the input trajectory why not put each of them into their own functions/file?
- Plot code is almost the same across all examples, create a general plot function?
- Remove the `robottrak_backup` file, its kind of redundant since we can use git to git it back

chutsu commented on matlab_simulation/code v1.0/ME597-7-Trajectories/robottraj.m:
```diff
@@ -1,85 +1,133 @@
+    w(2*c+1:5*c) = -2/5;
+
+    for t=1:length(T)
+        xddot(:,t) = [v(t)*cos(xd(3,t));
```
Can you not have used `motion_model()` function here? as well in lines 35, 57 and 80?

chutsu commented on matlab_simulation/code v1.0/ME597-7-Trajectories/motion_model.m:
```diff
@@ -0,0 +1,13 @@
+function xddot = motion_model(t, T, xd)
```
`v` and `w` both should be input variables to the motion model, also consider giving this function a better name other than `motion_model()`, I assume this is for a two wheeled robot?

chutsu commented on matlab_simulation/code v1.0/ME597-7-Trajectories/robottraj.m:
```diff
@@ -1,85 +1,133 @@
+    v = sin(0.2*T);
+    w = ones(size(T));
+    
+    for t=1:length(T)
```
The iteration of the motion model is very much the same across other examples, factor this for loop out? Same goes to all other examples?

neginlashkari commented:
This is a neat code to generate various trajectories for robots. This code allows the user to choose the function and its initial values. However, I think it would be better to instead of generally asking for initial conditions, let the user know what are the role of each initial variables in final trajectory. 
This way user can generate whatever trajectory he needs by entering initial conditions.

neginlashkari commented:
Also, when I choose "user defined motion model" as input trajectory, it gives me spiral function. I understood that we should write our desired function in "motion_mode.m". I was wondering if we can also make the user able to write his desired function in command window and get the results (instead of changing the mfile).

chutsu commented:
In response to allowing user to input desired function in the command window, I personally feel that would just add complexity to the simple example. You're already running matlab, its rather trivial to edit a file in the same directory, not to mention its rather annoying to type and correct mistakes in the command window multiple times.

yshiyi committed:
Delete extra file

yshiyi committed:
Replace with new file

yshiyi committed:
Replace with new file

yshiyi committed:
These are the newest version of the files.

Each individual motion has been generated in its own function. 
Users can select which trajectory to run and give any initial condition.
'motion_model.m' is the general motion model for all four pre-defined motions and also can be edited in each individual function.
Users are allowed to define and edit their own motion model in 'user_define_motion_model.m'. 
'input_check.m' can check whether the input index is valid or not.

yshiyi committed:
Add two motions

yshiyi committed:
Add two new motions

yshiyi committed:
Generate multiple nudges

yshiyi committed:
Add plots for two new motions

yshiyi committed:
Add two new motions and plot multiple nudges and swerves

yshiyi committed:
Add comments

yshiyi committed:
Update the available options

yshiyi commented:
This code only focused on generating different trajectories. Path planning is not considered.
Six different trajectories: spiral, nudges, swerves, corner, lane change, obstacle avoid are generated.
In addition, users can define their own trajectory.
Users are also allowed to give any initial condition except number 6, obstacle avoidance.
Each individual motion has been generated in its own function. 
Users can select which trajectory to run and give any initial condition as well.
'motion_model.m' is the general motion model for all six pre-defined motions and also can be edited in each individual function.
Users are allowed to define and edit their own motion model in 'user_define_motion_model.m'. 
'input_check.m' can check whether the input index is valid or not.
